File: frontend-dashboard/src/services/analytics.service.ts
import { BaseService } from "./base.service";
import {
  IChartData,
  IAnalyticsConfig,
} from "../models/interfaces/analytics.interface";
import { ANALYTICS_CONFIG } from "../utils/constants";
import { formatChartData } from "../utils/helpers";

export class AnalyticsService extends BaseService {
  private readonly config: IAnalyticsConfig;

  constructor() {
    super({ baseUrl: ANALYTICS_CONFIG.baseUrl });
    this.config = ANALYTICS_CONFIG;
  }

  // Inventory Analytics
  public async getStockLevelTrends(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.inventory.stockTrends
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getInventoryDistribution(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.inventory.distribution
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getWarehouseDistribution(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.inventory.warehouseDistribution
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getReorderPoints(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.inventory.reorderPoints
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  // Order Analytics
  public async getOrderFlow(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.orders.flow
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getOrderPipeline(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.orders.pipeline
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getOrderLifecycle(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.orders.lifecycle
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getOrderVolumeTrends(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.orders.volumeTrends
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getOrderValueDistribution(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.orders.valueDistribution
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getOrderPeakTimes(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.orders.peakTimes
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  // Transaction Analytics
  public async getTransactionVolume(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.transactions.volume
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getStockMovements(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.transactions.stockMovements
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getTransactionMetrics(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.transactions.metrics
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getTransactionPatterns(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.transactions.patterns
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getTransactionClusters(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.transactions.clusters
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getTransactionFlow(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.transactions.flow
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  // Performance Analytics
  public async getRealTimeMetrics(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.performance.metrics
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getPerformanceTrends(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.performance.trends
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getSystemHealth(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.performance.health
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getResourceUtilization(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.performance.resources
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  // Business Analytics
  public async getFinancialAnalytics(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.business.financial
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getRevenueAnalysis(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.business.revenue
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getHierarchicalData(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.business.hierarchical
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getForecastData(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.business.forecast
    );
    return formatChartData(response.data.data, response.data.metadata);
  }

  public async getTrendPredictions(): Promise<IChartData> {
    const response = await this.get<IChartData>(
      this.config.endpoints.business.trends
    );
    return formatChartData(response.data.data, response.data.metadata);
  }
}

File: logistics-engine/src/services/analytics_service.rs
use crate::{
    db::repository::analytics_repository::AnalyticsRepository,
    errors::{LogisticsError, Result},
    models::analytics::{
        FinancialAnalyticsResponse, InventoryDistributionResponse, OrderFlowResponse,
        OrderPipelineResponse, RealTimeMetricsResponse, ReorderPointsResponse,
        StockLevelTrendsResponse, TransactionVolumeResponse, WarehouseDistributionResponse,
    },
};
use std::sync::Arc;

#[derive(Clone)]
pub struct AnalyticsService {
    repository: Arc<AnalyticsRepository>,
}

impl AnalyticsService {
    pub fn new(repository: Arc<AnalyticsRepository>) -> Self {
        Self { repository }
    }

    // Inventory Analytics
    pub async fn get_stock_level_trends(&self) -> Result<StockLevelTrendsResponse> {
        let data = self.repository.get_stock_level_trends().await?;
        Ok(data)
    }

    pub async fn get_inventory_distribution(&self) -> Result<InventoryDistributionResponse> {
        let data = self.repository.get_inventory_distribution().await?;
        Ok(data)
    }

    pub async fn get_warehouse_distribution(&self) -> Result<WarehouseDistributionResponse> {
        let data = self.repository.get_warehouse_distribution().await?;
        Ok(data)
    }

    pub async fn get_reorder_points(&self) -> Result<ReorderPointsResponse> {
        let data = self.repository.get_reorder_points().await?;
        Ok(data)
    }

    // Order Analytics
    pub async fn get_order_flow(&self) -> Result<OrderFlowResponse> {
        let data = self.repository.get_order_flow().await?;
        Ok(data)
    }

    pub async fn get_order_pipeline(&self) -> Result<OrderPipelineResponse> {
        let data = self.repository.get_order_pipeline().await?;
        Ok(data)
    }

    pub async fn get_order_lifecycle(&self) -> Result<OrderPipelineResponse> {
        let data = self.repository.get_order_pipeline().await?;
        Ok(data)
    }

    pub async fn get_order_volume_trends(&self) -> Result<OrderFlowResponse> {
        let data = self.repository.get_order_flow().await?;
        Ok(data)
    }

    pub async fn get_order_value_distribution(&self) -> Result<OrderFlowResponse> {
        let data = self.repository.get_order_flow().await?;
        Ok(data)
    }

    pub async fn get_order_peak_times(&self) -> Result<OrderFlowResponse> {
        let data = self.repository.get_order_flow().await?;
        Ok(data)
    }

    // Transaction Analytics
    pub async fn get_transaction_volume(&self) -> Result<TransactionVolumeResponse> {
        let data = self.repository.get_transaction_volume().await?;
        Ok(data)
    }

    pub async fn get_stock_movements(&self) -> Result<TransactionVolumeResponse> {
        let data = self.repository.get_stock_movements().await?;
        Ok(data)
    }

    pub async fn get_transaction_metrics(&self) -> Result<TransactionVolumeResponse> {
        let data = self.repository.get_transaction_metrics().await?;
        Ok(data)
    }

    pub async fn get_transaction_patterns(&self) -> Result<TransactionVolumeResponse> {
        let data = self.repository.get_transaction_patterns().await?;
        Ok(data)
    }

    pub async fn get_transaction_clusters(&self) -> Result<TransactionVolumeResponse> {
        let data = self.repository.get_transaction_clusters().await?;
        Ok(data)
    }

    pub async fn get_transaction_flow(&self) -> Result<TransactionVolumeResponse> {
        let data = self.repository.get_transaction_flow().await?;
        Ok(data)
    }

    // Performance Analytics
    pub async fn get_real_time_metrics(&self) -> Result<RealTimeMetricsResponse> {
        let data = self.repository.get_real_time_metrics().await?;
        Ok(data)
    }

    pub async fn get_performance_trends(&self) -> Result<RealTimeMetricsResponse> {
        let data = self.repository.get_performance_trends().await?;
        Ok(data)
    }

    pub async fn get_system_health(&self) -> Result<RealTimeMetricsResponse> {
        let data = self.repository.get_system_health().await?;
        Ok(data)
    }

    pub async fn get_resource_utilization(&self) -> Result<RealTimeMetricsResponse> {
        let data = self.repository.get_resource_utilization().await?;
        Ok(data)
    }

    // Business Analytics
    pub async fn get_financial_analytics(&self) -> Result<FinancialAnalyticsResponse> {
        let data = self.repository.get_financial_analytics().await?;
        Ok(data)
    }

    pub async fn get_revenue_analysis(&self) -> Result<FinancialAnalyticsResponse> {
        let data = self.repository.get_revenue_analysis().await?;
        Ok(data)
    }

    pub async fn get_hierarchical_data(&self) -> Result<FinancialAnalyticsResponse> {
        let data = self.repository.get_hierarchical_data().await?;
        Ok(data)
    }

    pub async fn get_forecast_data(&self) -> Result<FinancialAnalyticsResponse> {
        let data = self.repository.get_forecast_data().await?;
        Ok(data)
    }

    pub async fn get_trend_predictions(&self) -> Result<FinancialAnalyticsResponse> {
        let data = self.repository.get_trend_predictions().await?;
        Ok(data)
    }
}
